api/v1alpha3/types.go

```diff
@@ -145,42 +140,48 @@ type IngressRule struct {
 // IngressRules is a slice of Azure ingress rules for security groups.
 type IngressRules []*IngressRule
 
-// PublicIP defines an Azure public IP address.
-type PublicIP struct {
-	ID        string `json:"id,omitempty"`
-	Name      string `json:"name,omitempty"`
-	IPAddress string `json:"ipAddress,omitempty"`
-	DNSName   string `json:"dnsName,omitempty"`
-}
-
 // LoadBalancer defines an Azure load balancer.
-type LoadBalancer struct {
-	ID               string           `json:"id,omitempty"`
-	Name             string           `json:"name,omitempty"`
-	SKU              SKU              `json:"sku,omitempty"`
-	FrontendIPConfig FrontendIPConfig `json:"frontendIpConfig,omitempty"`
-	BackendPool      BackendPool      `json:"backendPool,omitempty"`
-	Tags             Tags             `json:"tags,omitempty"`
+type LoadBalancerSpec struct {
+	ID          string       `json:"id,omitempty"`
+	Name        string       `json:"name,omitempty"`
+	SKU         SKU          `json:"sku,omitempty"`
+	FrontendIPs []FrontendIP `json:"frontendIPs,omitempty"`
+	Type        LBType       `json:"type,omitempty"`
 }
 
-// FrontendIPConfig - DO NOT USE
-// this empty struct is here to preserve backwards compatibility and should be removed in v1alpha4
-type FrontendIPConfig struct{}
-
 // SKU defines an Azure load balancer SKU.
 type SKU string
 
 const (
-	// SKUBasic is the value for the Azure load balancer Basic SKU
-	SKUBasic = SKU("Basic")
-	// SKUStandard is the value for the Azure load balancer Standard SKU
+	// SKUStandard is the value for the Azure load balancer Standard SKU.
 	SKUStandard = SKU("Standard")
 )
 
-// BackendPool defines a load balancer backend pool
-type BackendPool struct {
-	Name string `json:"name,omitempty"`
-	ID   string `json:"id,omitempty"`
+// Type defines an Azure load balancer Type.
+type LBType string
+
+const (
+	// Internal is the value for the Azure load b alancer internal type.
+	Internal = LBType("Internal")
+	// Public is the value for the Azure load balancer public type.
+	Public = LBType("Public")
+)
+
+// FrontendIP defines a load balancer frontend IP configuration.
+type FrontendIP struct {
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name"`
+	// +optional
+	PrivateIPAddress string `json:"privateIP,omitempty"`
+	// +optional
+	PublicIP *PublicIPSpec `json:"publicIP,omitempty"`
```

Comment:
@alexeldeib @MirzaSikander this supersedes #652, along with https://github.com/kubernetes-sigs/cluster-api/issues/3715